Add a blog post for the release of 2.2.0~beta2 (#98)
ocaml/platform-blog: opam-2-2-0-beta2.md

title: "opam 2.2.0 beta2 release"
authors: [
  "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(à)ocamlpro.com"}
  "Kate Deplaix - Ahrefs" {"mailto:kit-ty-kate(à)outlook.com"}
  "David Allsopp - Tarides" {"mailto:david(à)tarides.com"}
]
date: "2024-04-09"
--BODY--

<!---
_Feedback on this post is welcomed on [Discuss](https://discuss.ocaml.org/t/ann-opam-2-2-0-beta2/XXXX)!_
--->

> We are happy to announce the second beta release of opam 2.2.0.

Reviewer: ```suggestion
We are indescribably thrilled to announce the second beta release of opam 2.2.0.
```
🤪

It contains everything required to be able to make opam-repository compatible
with Windows, as well as a whole bunch of fixes. You can view the full list
of changes in the
[release note](https://github.com/ocaml/opam/releases/tag/2.2.0-beta2).

> **We'll post another blog post very soon with more direction on how to test

Reviewer: ```suggestion
**We'll post another blog post very soon with more directions on how to test
```

opam on Windows with this release.**

This version is a beta, we invite users to test it to spot previously
unnoticed bugs as we head towards the stable release.

## Changes

### Windows support

This beta introduces a handful of changes necessary to be able to
make the default opam-repository support Windows out of the box:

* Add a new `sys-ocaml-system` variable
* Hijack the `"%{?val_if_true:val_if_false:}%"` syntax to support extending

>   the variables of packages with + in their name

Reviewer: ```suggestion
  the variables of packages with + in their name (`conf-c++` and `conf-g++` already exist)
```

> * Add winsymlinks:native to the CYGWIN environment variable when installing
>   a package on Windows

Reviewer: ```suggestion
* Add `winsymlinks:native` to the CYGWIN environment variable when installing
  a package on Windows. In particular, this provides a workaround when extracting
  ocamlbuild's sources.
```

* Internal Cygwin installation's bin directory is placed as far down PATH as
  is necessary not to shadow `bash`, `tar`, `sort` or `git`
* Disable ACL in Cygwin internal install to avoid permission mismatch errors

We expect to be able to show the proposed changed to opam-repository
very soon to take advantage of all these changes.

> ### opam-repository scallability
> 
> The current draft resolution resulting from the discussion in
> [ocaml/opam-repository#23789](https://github.com/ocaml/opam-repository/issues/23789)
> about the scallability of opam-repository includes the removal of some

Reviewer: ```suggestion
### opam-repository scalability

The current draft resolution resulting from the discussion in
[ocaml/opam-repository#23789](https://github.com/ocaml/opam-repository/issues/23789)
about the scalability of opam-repository includes the removal of some
```

packages. However currently, opam uses the `patch` system command to apply
changes from a repository. The behaviour of that command is thus very
important and it is a known behaviour for the macOS and BSDs `patch`
command to not be able to delete files which leads to failures and
inconsistencies for opam. Package managers on those platforms installing

> opam already makes opam depend on GNU patch, however a certain number of

Reviewer: ```suggestion
opam already make opam depend on GNU patch, however a certain number of
```

> people do not install opam via a system package manager (e.g. our install script)

Reviewer: ```suggestion
people do not install opam via a system package manager (e.g. our own install script!)
```

and end up using their system version of `patch`. This is in particular
a problem on macOS as the name of the GNU patch command is not `gpatch` like

> on BSDs but simply `patch` when installed via Homebrew.

Author: I'm not a fan of this section but I haven't found a better way to describe the issue, but it's probably ok

Reviewer: I think it's OK!

To fix this issue, after many trials and errors

> (this issue being surprisingly tricky to fix), we've decided to:

Reviewer: Possible re-word:

This issue is surprisingly tricky to fix, and after many trials and errors,
we've decided to:

* Warn if GNU patch is not detected when a patch is applied
* Use `gpatch` by default instead of `patch` on NetBSD and DragonFlyBSD
* Use `gpatch` if it exists and is detected as GNU patch when `patch` is not
  GNU patch

These changes will make their way to the upcoming opam 2.1.6, in a few weeks.

### Other noteworthy changes

> * Recommend enabling Developer Mode on Windows

Reviewer: ```suggestion
* Recommend enabling Developer Mode on Windows. This allows the creation of symlinks without requiring elevation. Longer-term, the aim is that we should never _require_ Developer Mode, but at the moment more things work with it than without it!
```
(GitHub's rendering of the diff was awful when I tried doing this one big suggestion, so apologies for the string of individual comments!)

> * Fix MSYS2 support
> * Mark the internal cygwin installation as recommended

Reviewer: ```suggestion
* Mark the internal cygwin installation as recommended. Please don't try to maintain your own Cygwin install for use with opam unless you really know what you're doing!
* Fix MSYS2 support. For 2.2.0, the focus has been on Cygwin, so configuring opam to use MSYS2 is quite manual. Please don't try to maintain your own MSYS2 install for use with opam unless you _really really_ know what you're doing!
```

* Fix issues when using fish
* Improve the internal Cygwin installation during init on Windows
* Unixify Windows paths in init shells scripts

> * Disable Software Heritage fallbacks by default

Reviewer: Possibly worth a note as to why?

Author: Done in https://github.com/ocaml/platform-blog/pull/98/commits/817e9a05af87dd6d9ca2cd81ba247ee6e239a14d

* Make sure `opam source --dev` with git sources, clones the whole repository
  instead of using `--depth=1`
* Sandbox: Mark the user temporary directory
  (as returned by `getconf DARWIN_USER_TEMP_DIR`) as writable when TMPDIR
  is not defined on macOS

> * Add warning 69: Warn for new syntax when package name in variable in string
>   interpolation contains several '+'

Reviewer: ```suggestion
* Add warning 69: Warn for new syntax when package name in variable in string
  interpolation contains several '+' (this is related to the "hijack" item above)
```

> * Check for gpatch instead of patch on NetBSD and DragonFlyBSD

Reviewer: ```suggestion
```
(as mentioned previously?)

> * Add support for Wolfi OS, treat it like Apline family as it uses apk too

Reviewer: ```suggestion
* Add support for Wolfi OS, treat it like Alpine family as it also uses apk
```

> * Upgrade the vendored dune package to 3.14.2 to allow to compile opam when
>   the environment contains unicode characters on Windows

Reviewer: ```suggestion
* Upgrade the vendored dune package to 3.14.2 to allow to compile opam when
  the environment contains unicode characters on Windows (in particular, this means opam now works if your username contains accented characters)
```

* Upgrade other vendored packages

>   (cmdliner 1.2.0, re 1.11.0, ocamlgraph 2.1.0, opam-file-format 2.1.6)

Reviewer: FWIW the updates I don't think fix anything (in the way that the Dune one does) so I'm not sure it's worth putting?

Author: I think it's still worth knowing which versions of package they use are used in opam, in particular for the pre-built binaries. I did put it last because it's much less important but to me it *just* passes the threshold of "useful to know"

  "David Allsopp - Tarides" {"mailto:david(à)tarides.com"}
]
date: "2024-04-09"
--BODY--

<!---
_Feedback on this post is welcomed on [Discuss](https://discuss.ocaml.org/t/ann-opam-2-2-0-beta2/XXXX)!_
--->

We are indescribably thrilled to announce the second beta release of opam 2.2.0.

It contains everything required to be able to make opam-repository compatible
with Windows, as well as a whole bunch of fixes. You can view the full list
of changes in the
[release note](https://github.com/ocaml/opam/releases/tag/2.2.0-beta2).

**We'll post another blog post very soon with more directions on how to test
opam on Windows with this release.**

This version is a beta, we invite users to test it to spot previously
unnoticed bugs as we head towards the stable release.

## Changes

### Windows support

This beta introduces a handful of changes necessary to be able to
make the default opam-repository support Windows out of the box:

* Add a new `sys-ocaml-system` variable
* Hijack the `"%{?val_if_true:val_if_false:}%"` syntax to support extending
  the variables of packages with + in their name (`conf-c++` and `conf-g++` already exist)

> * Add `winsymlinks:native` to the CYGWIN environment variable when installing

Author: not this time ^^ the environment variable is usually all capitalized (https://www.cygwin.com/cygwin-ug-net/setup-env.html)

> * Add `winsymlinks:native` to the CYGWIN environment variable when installing

Reviewer 2: my mistake!
Then it is
```suggestion
* Add `winsymlinks:native` to the `CYGWIN` environment variable when installing
```

command to not be able to delete files which leads to failures and
inconsistencies for opam. Package managers on those platforms installing
opam already make opam depend on GNU patch, however a certain number of
people do not install opam via a system package manager (e.g. our own install script!)
and end up using their system version of `patch`. This is in particular
a problem on macOS as the name of the GNU patch command is not `gpatch` like
on BSDs but simply `patch` when installed via Homebrew.

This issue is surprisingly tricky to fix, and after many trials and errors,
we've decided to:

* Warn if GNU patch is not detected when a patch is applied
* Use `gpatch` by default instead of `patch` on NetBSD and DragonFlyBSD
* Use `gpatch` if it exists and is detected as GNU patch when `patch` is not
  GNU patch

These changes will make their way to the upcoming opam 2.1.6, in a few weeks.

### Other noteworthy changes

* Recommend enabling Developer Mode on Windows.
  This allows the creation of symlinks without requiring elevation.
  Longer-term, the aim is that we should never _require_ Developer Mode,
  but at the moment more things work with it than without it!

> * Mark the internal cygwin installation as recommended.

Reviewer 2: ```suggestion
* Mark the internal Cygwin installation as recommended.
```

  Please don't try to maintain your own Cygwin install
  for use with opam unless you really know what you're doing!
* Fix MSYS2 support. For 2.2.0, the focus has been on Cygwin,
  so configuring opam to use MSYS2 is quite manual.
  Please don't try to maintain your own MSYS2 install for use
  with opam unless you _really really_ know what you're doing!
* Fix issues when using fish
* Improve the internal Cygwin installation during init on Windows
* Unixify Windows paths in init shells scripts
* Disable Software Heritage fallbacks by default
* Make sure `opam source --dev` with git sources, clones the whole repository
  instead of using `--depth=1`
* Sandbox: Mark the user temporary directory
  (as returned by `getconf DARWIN_USER_TEMP_DIR`) as writable when TMPDIR
  is not defined on macOS

> * Add warning 69: Warn for new syntax when package name in variable in string

Reviewer 2: ```suggestion
* Add Warning 69: Warn for new syntax when package name in variable in string
```

title: "opam 2.2.0 beta2 release"
authors: [
  "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(à)ocamlpro.com"}
  "Kate Deplaix - Ahrefs" {"mailto:kit-ty-kate(à)outlook.com"}
  "David Allsopp - Tarides" {"mailto:david(à)tarides.com"}
]
date: "2024-04-09"
--BODY--

<!---
_Feedback on this post is welcomed on [Discuss](https://discuss.ocaml.org/t/ann-opam-2-2-0-beta2/XXXX)!_
--->

We are indescribably thrilled to announce the second beta release of opam 2.2.0.

It contains everything required to be able to make opam-repository compatible
with Windows, as well as a whole bunch of fixes. You can view the full list
of changes in the
[release note](https://github.com/ocaml/opam/releases/tag/2.2.0-beta2).

**We'll post another blog post very soon with more directions on how to test
opam on Windows with this release.**

This version is a beta, we invite users to test it to spot previously
unnoticed bugs as we head towards the stable release.

## Changes

### Windows support

This beta introduces a handful of changes necessary to be able to
make the default opam-repository support Windows out of the box:

> * Add a new `sys-ocaml-system` variable

Reviewer 2: ```suggestion
* Add a new `sys-ocaml-system` default opam variable
```

Reviewer: Or "global"?

Reviewer 2: default and global ?

Reviewer 2: ```suggestion
* Add a new `sys-ocaml-system` default global eval variable
```

a problem on macOS as the name of the GNU patch command is not `gpatch` like
on BSDs but simply `patch` when installed via Homebrew.

This issue is surprisingly tricky to fix, and after many trials and errors,
we've decided to:

* Warn if GNU patch is not detected when a patch is applied
* Use `gpatch` by default instead of `patch` on NetBSD and DragonFlyBSD
* Use `gpatch` if it exists and is detected as GNU patch when `patch` is not
  GNU patch

These changes will make their way to the upcoming opam 2.1.6, in a few weeks.

### Other noteworthy changes

* Recommend enabling Developer Mode on Windows.
  This allows the creation of symlinks without requiring elevation.
  Longer-term, the aim is that we should never _require_ Developer Mode,
  but at the moment more things work with it than without it!
* Mark the internal Cygwin installation as recommended.
  Please don't try to maintain your own Cygwin install
  for use with opam unless you really know what you're doing!
* Fix MSYS2 support. For 2.2.0, the focus has been on Cygwin,
  so configuring opam to use MSYS2 is quite manual.
  Please don't try to maintain your own MSYS2 install for use

>   with opam unless you _really really_ know what you're doing!

Reviewer 2: I don't understand that comment as opam don't have its own msys2 install

Reviewer: That why it's "your own". You can install MSYS2 (say to it's default `C:\msys64`) and then configure `opam` to use that at `opam init` - I'm merely suggesting that doing that is not for the faint-hearted/beginners (i.e. don't think "Oh, I have a Cygwin installation, opam can just use that" and definitely don't think "Oh, I have an MSYS2 installation, opam can just use that, it'll be faster and everything")

* Unixify Windows paths in init shells scripts
* Disable Software Heritage fallback by default as there currently no CI job
  in opam-repository to check validity of proposed `swhid` regarding release
  archive (neither publication tools support) and some concerns were raised
  regarding the degree of trust of the hashing method used by Software
  Heritage (sha1).
* Make sure `opam source --dev` with git sources, clones the whole repository
  instead of using `--depth=1`
* Sandbox: Mark the user temporary directory
  (as returned by `getconf DARWIN_USER_TEMP_DIR`) as writable when TMPDIR
  is not defined on macOS
* Add Warning 69: Warn for new syntax when package name in variable in string
  interpolation contains several '+' (this is related to the "hijack" item above)
* Add support for Wolfi OS, treat it like Alpine family as it also uses apk
* Upgrade the vendored dune package to 3.14.2 to allow to compile opam when
  the environment contains unicode characters on Windows (in particular,
  this means opam now works if your username contains accented characters)
* Upgrade other vendored packages
  (cmdliner 1.2.0, re 1.11.0, ocamlgraph 2.1.0, opam-file-format 2.1.6)

> Various other improvements were made and bugs were fixed.

Reviewer 2: ```suggestion
Various other general and performance improvements were made and bugs were fixed.
```
